scipy/special/_basic.py

```diff
@@ -2065,14 +2072,27 @@ def factorial(n, exact=False):
                     current = un[i + 1]
                     val *= _range_prod(prev, current)
                     out[n == current] = val
+
+            if np.isnan(n).any():
+                out = out.astype(np.float64)
+                out[np.isnan(n)] = np.nan
             return out
     else:
         if np.ndim(n) == 0:
-            return 0 if n < 0 else gamma(n + 1)
+            if not np.isnan(n):
+                return 0 if n < 0 else gamma(n + 1)
+            else:
+                return np.nan
 
         n = asarray(n)
         vals = gamma(n + 1)
-        return where(n >= 0, vals, 0)
+        # Ignore runtime warning when less operator used w/np.nan
```

Comment:
We can simplify here by reworking the logic a bit-do something like

```python
out = np.empty_like(n)
with np.errstate(all='ignore'):
    neg = n < 0
out[neg] = 0
out[~neg] = gamma(n + 1)
```

Since `gamma` will propagate NaNs correctly we don't have to worry about that part.

Even better would be to make a tiny private ufunc with a kernel like:

```cython
cdef double factorial(double n) nogil:
    if n < 0:
        return 0
    else:
        return gamma(n + 1)
```

and then this entire `else` branch can just be calling that function.

Reply:
@person142
I like the private ufuncs idea plus that may be a useful pattern to follow in future maintenance work in special. 

Should the ufunc go under cephes or somewhere else? (if somewhere else I'm not sure where)

Reply:
I would default to Cython (we have a vague hope that Cython is more maintainable for people that don't know C) and currently we just put all Cython stuff in the `special` directory, something like `special/_factorial.pxd`. You can get at `gamma` by doing a

```cython
from ._cephes cimport gamma
```

I imagine that ultimately most of the factorial implementation would benefit from being rewritten in Cython, so this would be a good start for that too.

Reply:
@person142 I started doing this and ran into some build failures, not sure if i'm missing something. I was cribbing off this PR: 
https://github.com/scipy/scipy/pull/10440/files
does that look like what you had in mind (modulo the changes to be fore factorial)? 

I though I was, but maybeI'm still not clear on the ***full*** build process for `scipy.special`.

Reply:
Yeah, that PR should be fairly representative. You can always push a partial commit if further debugging is needed.

Reply:
I figured it out, I was missing the empty, "internal do not use" docs which cause the build to fail. After that and a couple other minor fails I was able to rebuild and do some fixups/cleanups w/`_factorial`.